proposals/0000-custom-executors.md

```diff
@@ -0,0 +1,464 @@
+concurrency.
+
+### Executors and jobs
+
+All executors must conform to the following protocol:
+
+```swift
+protocol Executor: AnyObject {
+  /// Enqueue a job on this executor to run asynchronously.
+  func enqueue(_ job: UnownedJobRef)
+
+  /// Get an unowned reference to this executor. The reference
+  /// must remain valid as long as the executor is.
+  func asUnownedRef() -> UnownedExecutorRef
+}
+```
+
+Executors are required to follow certain ordering rules when executing
+their jobs:
+
+- The call to `job.execute(currentExecutor:)` must happen-after the call
+  to `enqueue(_:)`.
+
+- If the executor is a serial executor, then the execution of all jobs
+  must be totally ordered: for any two different jobs *A* and *B*
+  submitted to the same executor with `enqueue(_:)`, it must be true
+  that either all events in *A* happen-before all events in *B* or all
+  events in *B* happen-before all events in *A*.
+
+The `UnownedJobRef` type is the opaque type of schedulable jobs.
+The job reference is "self-owning", meaning that executing it
+takes over ownership of the reference, potentially invalidating
+it immediately. The executor must not assume the validity of the
+job after executing it. (If Swift supported move-only types,
+`UnownedJobRef` could be one, and `execute(currentExecutor:)` would
+be a consuming method.)
+
+```swift
+struct UnownedJobRef {
+  /// Get the requested priority of the job.
+  var priority: Priority { get }
+
+  /// Execute the job on the current thread, claiming to be
+  /// running on the given executor. The executor reference must
+  /// remain valid during this call unless it is a serial executor
+  /// which the job successfully gives up.
+  ///
+  /// Calling this immediately invalidates the job reference
+  /// from the caller's perspective, and the caller must not
+  /// refer to the job again.
+  func execute(currentExecutor: UnownedExecutorRef)
+
+  /// Create an unsafe job reference to run the given function
+  /// at the given priority. This may require allocating memory.
+  init(priority: Priority = .default, operation: @escaping () -> ())
+}
+```
+
+The `UnownedExecutorRef` type is the opaque type of a reference
+to an executor. This type packs certain highly valuable information
+into the reference. It is an unmanaged (that is, unsafe) reference
+to the executor. Whatever context produces an `UnownedExecutorRef`
+is generally responsible for keeping the reference alive while it
+is in use, often by maintaining some other stable relationship.
+
+The identity of an `UnownedExecutorRef` is determined by object
+identity.  The appropriate flags must be set consistently for the
+executor.
+
+```swift
+struct UnownedExecutorRef: Equatable {
+  init<T: SerialExecutor>(serialExecutor: T, supportsSwitching: Bool)
+  init<T: Executor>(concurrentExecutor: T)
+
+  static var defaultConcurrent: UnownedExecutorRef { get }
+
+  var isSerial: Bool { get }
+
+  func asOwned() -> Executor
```

Comment:
So the unowned is to avoid refcounting, and this vends out the "actual" one right?

Reply:
It's a way to convert back to the original executor in Swift code, yeah.